### PR TITLE
Support for credentials in form body for /token endpoint and make state parameter optional

### PR DESCRIPTION
**Pull request by csantos**
## Description
Added support for credentials in the form body (in addition to previous Authorization header support) to the /token endpoint. Removed requirement that the state parameter from the /authorize endpoint. These changes allow openidconnect.net and Apigee's integrated portal to work with Identity Facade without any additional work.

- Added new EV-ClientCredentialsFromBody step to extract credentials from from body
- Conditionally executing steps to either extract credentials from header or form body
- Removed step that conditionally executed RF-InvalidRequest-Redirect if state parameter wasn't provided.

## Housekeeping
(please check all that apply [x], do not edit the text)
- [X] I have run all the tests locally and they all pass.
- [X] I have followed the relevant style guide for my changes.

## Full Repo Validation Required
(please check all that apply [x], do not edit the text)
- [ ] PR requires full pipeline run (Run for changes only by default).

**CC:** @apigee-devrel-reviewers

### CODE REVIEW

**Comment by OmidTahouri** [comment]
Hi @csantos -- thanks for your contribution!

Although the `state` parameter is technically optional as per the OAuth specification, it is generally highly recommended and is therefore a hard dependancy for the Identity Facade pattern implemented here. The proxy leverages `state` to capture all the parameters on the `/authorize` request so that they can be extracted and associated with the subsequent `/token` request. In a way, the `state` is used to maintain a session (or _state_) across requests. Without this, I can't see how the `/token` endpoint can be sure that the same client came from the corresponding `/authorize` endpoint where the `client_id`, `redirect_uri`, etc. were validated.

I'll let @JoelGauci chime in here as well, to validate my thinking...
I'm sure we can discuss this over a short call if it's easier.

Thanks again!

**Comment by csantos** [comment]
Thanks for the quick response @OmidTahouri. Yep, totally agree that using the state parameter is to mitigate CSRF attacks, and that using it is a best practice. It is an optional param though, and the Integrated Portal does not pass it which hampers the ability to use the "Try this API" functionality (or any other client that doesn't pass this param) w/Identity Facade. 

Identity Facade does use state2 for the proxy to IdP calls, making those calls more secure...so it seems it is more about whether to enforce that best practice and Identity Facade taking the stance that it is required?

Let me know if you'd like me to not include making state optional. I'll rollback that change and only have the added support for credentials in form body for /token endpoint. If so, did you want me to open up a separate pull request? That way I can clean up the history a bit and not see removal / putting back the state param requirement?

**Comment by JoelGauci** [comment]
According to me it is ok to make the state parameter optional, but I think we should in this case highly recommend and propose it by default in the proxy configuration.

In the logic that is implemented, the state parameter (state1) is retrieved from the original request (/authorize) and stored into a "session" (state2).

state2 is used by Apigee to redirect the end-user with the identity provider (IdP)
Then state1 is sent back to the client app at the end of the /callback processing.

As you mention Carlos, we can make it work without state1 and we will always implement the state2 used in the relation between Apigee and the IdP.

**Comment by OmidTahouri** [comment]
Thanks for clarifying, @JoelGauci

To summarise, it's imperative we maintain a state parameter between the proxy and the IdP, but not between the client and proxy.

**Review by JoelGauci — COMMENTED** [review]
Carlos,

- ok for me on the configuration of policies and proxy

- could you modify the README to clearly specify that we do recommend to use the state paramater to mitigate risk of attacks even though it is optional?

Thanks

**Comment by csantos** [comment]
@JoelGauci no problem. I've taken a stab at the update. Any feedback is welcomed.

@OmidTahouri thanks for bringing this up as a topic of discussion.

File: references/identity-facade/apiproxy/policies/EV-ClientCredentialsFromBody.xml [hunk 1]
```diff
@@ -0,0 +1,11 @@
+<?xml version="1.0" encoding="UTF-8" standalone="yes"?>
```
Comment (danistrebel) [inline references/identity-facade/apiproxy/policies/EV-ClientCredentialsFromBody.xml:1]: Could you please add the default license header here?

**Review by JoelGauci — APPROVED** [review]

File: references/identity-facade/apiproxy/policies/EV-ClientCredentialsFromBody.xml (hunk 1 above)
Comment (csantos) [reply references/identity-facade/apiproxy/policies/EV-ClientCredentialsFromBody.xml:1]: Doh! Done. Totally missed that.

**Pull request merged**